kernel/src/hil/usb_hid.rs

```diff
@@ -0,0 +1,86 @@
+//! Interface for USB HID (Human Interface Device) class
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// Called when a packet is received.
+    /// This will return the buffer passed into `receive_buffer()` as well as
+    /// the endpoint where the data was received. The `len` argument contains
+    /// the length of the packet received. If the buffer length is smaller
+    /// then this length the buffer will only contain part of the packet and
+    /// `result` will contain indicate an `ESIZE` error.
+    fn packet_received(
+        &'a self,
+        result: ReturnCode,
+        buffer: &'static mut [u8; 64],
+        len: usize,
+        endpoint: usize,
+    );
+
+    /// Called when a packet has been finished transmitting.
+    /// This will return the buffer passed into `send_buffer()` as well as
+    /// the endpoint where the data was sent. The `len` argument contains
+    /// the number of bytes sent. If this is less then the size of the
+    /// buffer `result` will contain the `ESIZE` error.
+    fn packet_transmitted(
+        &'a self,
+        result: ReturnCode,
+        buffer: &'static mut [u8; 64],
+        len: usize,
+        endpoint: usize,
+    );
+
+    /// Called when checking if we can start a new receive operation.
+    /// Should return true if we are ready to receive and not currently
+    /// in the process of receiving anything. That is if we are currently
+    /// idle.
+    /// If there is an outstanding call to receive, a callback already
+    /// waiting to be called then this will return false.
+    fn can_receive(&'a self) -> bool;
+}
+
+pub trait UsbHid<'a> {
+    /// Sets the buffer to be sent and starts a send transaction.
+    /// On success returns the number of bytes that will be sent.
+    /// On failure returns an error code and the buffer passed in.
+    fn send_buffer(
+        &'a self,
+        send: &'static mut [u8; 64],
+    ) -> Result<usize, (ReturnCode, &'static mut [u8; 64])>;
+
+    /// Cancels a send called by `send_buffer()`. Once called the
+    /// `packet_transmitted()` callback won't occur and no new data
+    /// will be sent. Any data currently being sent will stil be sent.
+    /// On success (a send was cancelled) returns the buffer passed to
+    /// `send_buffer()`.
+    /// On failure returns a return code
+    fn send_cancel(&'a self) -> Result<&'static mut [u8; 64], ReturnCode>;
```

Comment:
I recall @phil-levis thinking pretty carefully through the async cancellation semantics elsewhere (I believe in the radio stack?). I think it would be beneficial to spell out the cancellation semantics a bit more explicitly so as to make sure that when cancellation occurs, it's clearly deterministic to the caller what will happen.

I believe the radio semantics were that either (a) cancellation failed and a callback will still come, OR (b) cancellation succeeded and no callback will come.

The "any data currently being sent will still be sent" part here reads a bit ambiguously to me -- if I successfully cancel "more" (?) data from being sent, will I get both a successful cancel and a send-complete callback?

Reply:
@ppannuto You have the semantics right: SUCCESS means there will be no callback, an error code means there will be (and the callback may return a SUCCESS or an ECANCEL).

Reply:
Note one issue here: what if there is no call outstanding? You don't want to return a buffer and you don't want to signal a callback. This is why the radio moved towards cancel basically always invoking a callback: SUCCESS meant there was no outstanding call. This also simplified buffer management as there's a single call path to get the buffer back.

Reply:
I think the following semantics would make sense:
- `Ok(buffer)` => A buffer was already set, and could successfully be cancelled.
- `Err(ECANCEL)` => A buffer was already set, but couldn't be cancelled as the send operation is in progress. A callback returning the buffer will be issued later.
- `Err(EALREADY)` => No buffer is set at the moment (i.e. either no buffer was ever set, or the previously set buffer was already returned by a callback or cancel operation). I think `EALREADY` is the closest error code among what we already have (i.e. "already cancelled").

Reply:
> SUCCESS meant there was no outstanding call.

I'm not sure it's better than `EALREADY`, because even though the result is a success (nothing is left pending), no "cancelling" happened. If the caller expects to get back a buffer at this point (likely due to an invalid state tracking by the caller - i.e. a bug in the caller's code), they cannot just ignore a "success".